File: Roadmap/31 - SIMULADOR JUEGOS OLÍMPICOS/c#/sisaroot.cs
// #31 SIMULADOR JUEGOS OLÍMPICOS - SisaRoot

using System;
using System.Collections.Generic;
using System.Linq;

class Participant
{
    public string Name { get; set; }
    public string Country { get; set; }
    public Participant(string name, string country) { Name = name; Country = country; }
}

class OlympicEvent
{
    public string Name { get; set; }
    public List<Participant> Participants { get; set; } = new();
    public OlympicEvent(string name) => Name = name;
}

class CountryMedals
{
    public string Country { get; set; }
    public int Gold { get; set; }
    public int Silver { get; set; }
    public int Bronze { get; set; }
    public int Total => Gold + Silver + Bronze;
    public CountryMedals(string country) => Country = country;
}

class SisaRoot
{
    static readonly List<OlympicEvent> events = new();
    static readonly Dictionary<string, CountryMedals> medalTable = new();
    static readonly Random rng = new();

    static void RegisterEvent()
    {
        Console.Write("Nombre del evento: ");
        string name = Console.ReadLine()?.Trim() ?? "";
        if (string.IsNullOrEmpty(name)) { Console.WriteLine("Nombre inválido."); return; }
        if (events.Any(e => e.Name.Equals(name, StringComparison.OrdinalIgnoreCase)))
        { Console.WriteLine($"El evento '{name}' ya existe."); return; }
        events.Add(new OlympicEvent(name));
        Console.WriteLine($"Evento '{name}' registrado.");
    }

    static void RegisterParticipant()
    {
        if (events.Count == 0) { Console.WriteLine("No hay eventos registrados."); return; }
        Console.WriteLine("Eventos disponibles:");
        for (int i = 0; i < events.Count; i++)
            Console.WriteLine($"  {i + 1}. {events[i].Name}");
        Console.Write("Selecciona el número del evento: ");
        if (!int.TryParse(Console.ReadLine(), out int idx) || idx < 1 || idx > events.Count)
        { Console.WriteLine("Selección inválida."); return; }
        var ev = events[idx - 1];
        Console.Write("Nombre del participante: ");
        string pname = Console.ReadLine()?.Trim() ?? "";
        Console.Write("País del participante: ");
        string country = Console.ReadLine()?.Trim() ?? "";
        if (string.IsNullOrEmpty(pname) || string.IsNullOrEmpty(country))
        { Console.WriteLine("Datos inválidos."); return; }
        ev.Participants.Add(new Participant(pname, country));
        Console.WriteLine($"Participante '{pname} ({country})' añadido a '{ev.Name}'.");
    }

    static void AddMedal(string country, string type)
    {
        if (!medalTable.TryGetValue(country, out var cm))
        { cm = new CountryMedals(country); medalTable[country] = cm; }
        if (type == "gold") cm.Gold++;
        else if (type == "silver") cm.Silver++;
        else cm.Bronze++;
    }

    static void SimulateEvents()
    {
        if (events.Count == 0) { Console.WriteLine("No hay eventos para simular."); return; }
        foreach (var ev in events)
        {
            Console.WriteLine($"\n=== Simulando: {ev.Name} ===");
            if (ev.Participants.Count < 3)
            { Console.WriteLine($"  Necesita >=3 participantes (tiene {ev.Participants.Count}). Saltando."); continue; }
            var shuffled = ev.Participants.OrderBy(_ => rng.Next()).ToList();
            var gold = shuffled[0]; var silver = shuffled[1]; var bronze = shuffled[2];
            Console.WriteLine($"  🥇 Oro:    {gold.Name} ({gold.Country})");
            Console.WriteLine($"  🥈 Plata:  {silver.Name} ({silver.Country})");
            Console.WriteLine($"  🥉 Bronce: {bronze.Name} ({bronze.Country})");
            AddMedal(gold.Country, "gold"); AddMedal(silver.Country, "silver"); AddMedal(bronze.Country, "bronze");
        }
    }

    static void ShowReport()
    {
        Console.WriteLine("\n==============================");
        Console.WriteLine("   INFORME FINAL - JJOO 2024  ");
        Console.WriteLine("==============================");
        if (medalTable.Count == 0) { Console.WriteLine("No se han simulado eventos aún."); return; }
        var ranking = medalTable.Values
            .OrderByDescending(c => c.Gold).ThenByDescending(c => c.Silver).ThenByDescending(c => c.Bronze).ToList();
        Console.WriteLine($"{"Pos.",-5} {"País",-25} {"Oro",-6} {"Plata",-6} {"Bronce",-6} {"Total",-6}");
        Console.WriteLine(new string('-', 60));
        for (int i = 0; i < ranking.Count; i++)
        {
            var c = ranking[i];
            Console.WriteLine($"{i + 1 + ".",-5} {c.Country,-25} {c.Gold,-6} {c.Silver,-6} {c.Bronze,-6} {c.Total,-6}");
        }
    }

    static void Main()
    {
        while (true)
        {
            Console.WriteLine("\n====== SIMULADOR JJOO PARÍS 2024 ======");
            Console.WriteLine("1. Registrar evento\n2. Registrar participante\n3. Simular eventos\n4. Ver informe\n5. Salir");
            Console.Write("Selecciona una opción: ");
            switch (Console.ReadLine()?.Trim())
            {
                case "1": RegisterEvent(); break;
                case "2": RegisterParticipant(); break;
                case "3": SimulateEvents(); break;
                case "4": ShowReport(); break;
                case "5": Console.WriteLine("¡Hasta luego!"); return;
                default: Console.WriteLine("Opción no válida."); break;
            }
        }
    }
}

File: Roadmap/31 - SIMULADOR JUEGOS OLÍMPICOS/python/sisaroot.py
# #31 SIMULADOR JUEGOS OLÍMPICOS - SisaRoot

import random
from collections import defaultdict


def shuffle_participants(participants: list) -> list:
    shuffled = participants[:]
    random.shuffle(shuffled)
    return shuffled


class OlympicSimulator:
    def __init__(self):
        self.events: dict[str, list[dict]] = {}
        self.medal_table: dict[str, dict] = defaultdict(lambda: {"gold": 0, "silver": 0, "bronze": 0})
        self.results: list[dict] = []

    def register_event(self):
        name = input("Nombre del evento: ").strip()
        if not name:
            print("Nombre inválido.")
            return
        if name.lower() in (k.lower() for k in self.events):
            print(f"El evento '{name}' ya existe.")
            return
        self.events[name] = []
        print(f"Evento '{name}' registrado.")

    def register_participant(self):
        if not self.events:
            print("No hay eventos registrados. Registra un evento primero.")
            return
        print("Eventos disponibles:")
        event_names = list(self.events.keys())
        for i, ev in enumerate(event_names, 1):
            print(f"  {i}. {ev}")
        try:
            idx = int(input("Selecciona el número del evento: ").strip())
            if not (1 <= idx <= len(event_names)):
                raise ValueError
        except ValueError:
            print("Selección inválida.")
            return
        event_name = event_names[idx - 1]
        participant_name = input("Nombre del participante: ").strip()
        country = input("País del participante: ").strip()
        if not participant_name or not country:
            print("Datos inválidos.")
            return
        self.events[event_name].append({"name": participant_name, "country": country})
        print(f"Participante '{participant_name} ({country})' registrado en '{event_name}'.")

    def simulate_events(self):
        if not self.events:
            print("No hay eventos para simular.")
            return
        self.results.clear()
        for event_name, participants in self.events.items():
            print(f"\n=== Simulando: {event_name} ===")
            if len(participants) < 3:
                print(f"  Necesita al menos 3 participantes (tiene {len(participants)}). Saltando.")
                continue
            shuffled = shuffle_participants(participants)
            gold   = shuffled[0]
            silver = shuffled[1]
            bronze = shuffled[2]
            print(f"  🥇 Oro:    {gold['name']} ({gold['country']})")
            print(f"  🥈 Plata:  {silver['name']} ({silver['country']})")
            print(f"  🥉 Bronce: {bronze['name']} ({bronze['country']})")
            self.medal_table[gold["country"]]["gold"] += 1
            self.medal_table[silver["country"]]["silver"] += 1
            self.medal_table[bronze["country"]]["bronze"] += 1
            self.results.append({
                "event": event_name,
                "gold": gold, "silver": silver, "bronze": bronze,
            })

    def show_report(self):
        print("\n" + "=" * 50)
        print("        INFORME FINAL — JJOO PARÍS 2024")
        print("=" * 50)
        if not self.results:
            print("No se han simulado eventos aún.")
            return
        print("\n📋 GANADORES POR EVENTO:")
        for result in self.results:
            print(f"\n  Evento: {result['event']}")
            print(f"    🥇 {result['gold']['name']} ({result['gold']['country']})")
            print(f"    🥈 {result['silver']['name']} ({result['silver']['country']})")
            print(f"    🥉 {result['bronze']['name']} ({result['bronze']['country']})")
        print("\n🏆 RANKING DE PAÍSES:")
        ranking = sorted(
            self.medal_table.items(),
            key=lambda x: (-x[1]["gold"], -x[1]["silver"], -x[1]["bronze"])
        )
        print(f"{'Pos.':<5} {'País':<25} {'🥇':<6} {'🥈':<6} {'🥉':<6} {'Total':<6}")
        print("-" * 58)
        for pos, (country, medals) in enumerate(ranking, 1):
            total = medals["gold"] + medals["silver"] + medals["bronze"]
            print(f"{pos:<5} {country:<25} {medals['gold']:<6} {medals['silver']:<6} {medals['bronze']:<6} {total:<6}")

    def run(self):
        menu = {
            "1": ("Registrar evento",       self.register_event),
            "2": ("Registrar participante", self.register_participant),
            "3": ("Simular eventos",        self.simulate_events),
            "4": ("Ver informe final",      self.show_report),
            "5": ("Salir",                  None),
        }
        while True:
            print("\n====== SIMULADOR JJOO PARÍS 2024 ======")
            for key, (label, _) in menu.items():
                print(f"{key}. {label}")
            option = input("Selecciona una opción: ").strip()
            if option == "5":
                print("¡Hasta luego!")
                break
            elif option in menu:
                menu[option][1]()
            else:
                print("Opción no válida.")


if __name__ == "__main__":
    OlympicSimulator().run()
